File: 229-majority-element-ii/majority-element-ii.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        
        int count1=1,count2=0,num1=nums[0],num2=nums[0];

        for(int i=1;i<nums.size();i++){

            if(nums[i]==num1){
                count1++;
            }
            else if(nums[i]==num2){
                count2++;
            }
            else if(count1==0){
                num1 = nums[i];
                count1++;
            }
            else if(count2==0){
                num2 = nums[i];
                count2++;
            }
            else{
                count1--;count2--;
            }
        }
        count1=0;count2=0;
        for(int i=0;i<nums.size();i++){
            if(nums[i]==num1){
                count1++;
            }else if(nums[i]==num2){
                count2++;
            }
        }
        vector<int> res;
        int k = nums.size()/3;
        if(count1>k){
            res.push_back(num1);
        }
        if(count2>k){
            res.push_back(num2);
        }
        return res;
    }
};

File: 229-majority-element-ii/majority-element-ii.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        List<Integer> li = new ArrayList<>();
        int[] count = new int[2];
        int[] arr = new int[2];
        arr[0]=nums[0];
        arr[1]=nums[0];
        count[0]=1;
        
        for(int i=1;i<nums.length;i++){
            if(nums[i]==arr[0]){
                count[0]++;
            }else if(nums[i]==arr[1]){
                count[1]++;
            }else if(count[0]==0){
               arr[0]=nums[i];
               count[0]++;
            }else if(count[1]==0){
                arr[1]=nums[i];
                count[1]++;
            }else{
                count[0]--;
                count[1]--;
            }
        }
        count[0]=0;
        count[1]=0;
        for(int i=0;i<nums.length;i++){
            if(arr[0]==nums[i]){
                count[0]++;
            }else if(arr[1]==nums[i]){
                count[1]++;
            }
        }
        int k = nums.length/3;
        if(count[0]>k){
            li.add(arr[0]);
        }
        if(count[1]>k){
            li.add(arr[1]);
        }
        return li;
    }
}